From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RuscalWorld <me@ruscalworld.ru>
Date: Fri, 4 Mar 2022 19:46:51 +0300
Subject: [PATCH] Masquerade protected commands


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b4493c0622f63a579527bcff1cbd26038fff3b5a..fa106f48271df39401adeb2ee68cde00202f3400 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2811,6 +2811,8 @@ public final class CraftServer implements Server {
 
     @Override
     public String getPermissionMessage() {
+        if (ru.bortexel.utopia.UtopiaConfig.masqueradeProtectedCommands)
+            return org.spigotmc.SpigotConfig.unknownCommandMessage;
         return com.destroystokyo.paper.PaperConfig.noPermissionMessage;
     }
 
diff --git a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
index 58e11ac13d95b6c506cbdf150a010708dc6c2d3c..59e5b3e143fcce7d5ba8aec293e4721b011f2a5b 100644
--- a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
+++ b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
@@ -126,4 +126,10 @@ public class UtopiaConfig {
     private static void loadControlledProperty(ControlledProperty property) {
         controlledProperties.put(property.getName(), property);
     }
+
+    public static String STUFF_ROOT = "stuff";
+    public static boolean masqueradeProtectedCommands = false;
+    private static void configStuff() {
+        masqueradeProtectedCommands = getBoolean(STUFF_ROOT + ".masquerade-protected-commands", false);
+    }
 }
