From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RuscalWorld <me@ruscalworld.ru>
Date: Sun, 12 Jun 2022 16:27:41 +0300
Subject: [PATCH] Masquerade protected commands


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9a75bad868752cad67288d8931156fe60452418b..e91b8aeaf75d3ef007df28d7bc27aed300bf9dcc 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2883,6 +2883,10 @@ public final class CraftServer implements Server {
 
     @Override
     public String getPermissionMessage() {
+        // Utopia start
+        if (ru.bortexel.utopia.UtopiaConfig.masqueradeProtectedCommands)
+            return org.spigotmc.SpigotConfig.unknownCommandMessage;
+        // Utopia end
         return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacyAmpersand().serialize(io.papermc.paper.configuration.GlobalConfiguration.get().messages.noPermission);
     }
 
diff --git a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
index b6afe711cad47e34280b7b44e363ca770c1494b3..66808fd60b1e01d31f62b7bab844e5f17887522c 100644
--- a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
+++ b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
@@ -128,4 +128,10 @@ public class UtopiaConfig {
     private static void loadControlledProperty(ControlledProperty property) {
         controlledProperties.put(property.getName(), property);
     }
+
+    public static String STUFF_ROOT = "stuff";
+    public static boolean masqueradeProtectedCommands = false;
+    private static void configStuff() {
+        masqueradeProtectedCommands = getBoolean(STUFF_ROOT + ".masquerade-protected-commands", false);
+    }
 }
