From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruscalworld <me@ruscal.world>
Date: Thu, 9 Mar 2023 10:58:36 +0300
Subject: [PATCH] Masquerade protected commands


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 799a9e037729a579351266ab6aa5ea54bcf18447..dbb571bc95c6ae54a2b87bfb142e0643bb5553ee 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2849,6 +2849,10 @@ public final class CraftServer implements Server {
 
     @Override
     public String getPermissionMessage() {
+        // Utopia start
+        if (ru.bortexel.utopia.UtopiaConfig.masqueradeProtectedCommands)
+            return org.spigotmc.SpigotConfig.unknownCommandMessage;
+        // Utopia end
         return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacyAmpersand().serialize(io.papermc.paper.configuration.GlobalConfiguration.get().messages.noPermission);
     }
 
diff --git a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
index 3267488f2621d386e1662a449dcda7aa7c3a1142..6d6bf162e7fb78159522b771b022b95633d3ece5 100644
--- a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
+++ b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
@@ -129,4 +129,10 @@ public class UtopiaConfig {
     private static void loadControlledProperty(ControlledProperty property) {
         controlledProperties.put(property.getName(), property);
     }
+
+    public static String STUFF_ROOT = "stuff";
+    public static boolean masqueradeProtectedCommands = false;
+    private static void configStuff() {
+        masqueradeProtectedCommands = getBoolean(STUFF_ROOT + ".masquerade-protected-commands", false);
+    }
 }
