From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RuscalWorld <me@ruscalworld.ru>
Date: Thu, 4 Aug 2022 07:58:26 +0300
Subject: [PATCH] Implement performance watchdog


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 519883c5549744e047a8a96afee142746fcb49cd..fa7132be39fde2bdaed1ee5936ec387e20bfae88 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -4,10 +4,8 @@ import com.google.common.collect.Lists;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
-import java.io.BufferedReader;
-import java.io.BufferedWriter;
-import java.io.IOException;
-import java.io.InputStreamReader;
+
+import java.io.*;
 import java.net.InetAddress;
 import java.net.Proxy;
 import java.nio.charset.StandardCharsets;
@@ -229,6 +227,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         gg.pufferfish.pufferfish.PufferfishConfig.load(); // Pufferfish
         gg.pufferfish.pufferfish.PufferfishCommand.init(); // Pufferfish
 
+        ru.bortexel.utopia.UtopiaConfig.init((File) options.valueOf("utopia-settings"));
+
         this.setPvpAllowed(dedicatedserverproperties.pvp);
         this.setFlightAllowed(dedicatedserverproperties.allowFlight);
         this.setMotd(dedicatedserverproperties.motd);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index eb2a1c5c98c73d232439aa1bf766ead17a65d5d3..61a7f16d0ff2387069163e5fa0937d742670e060 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -25,21 +25,7 @@ import java.io.IOException;
 import java.io.InputStreamReader;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
-import java.util.Base64;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.LinkedHashMap;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
 import java.util.function.Consumer;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -243,6 +229,9 @@ import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.MarkedYAMLException;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
+import ru.bortexel.utopia.UtopiaConfig;
+import ru.bortexel.utopia.management.PerformanceWatchdog;
+import ru.bortexel.utopia.management.PropertyPool;
 
 import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
@@ -372,6 +361,12 @@ public final class CraftServer implements Server {
             MapPalette.setMapColorCache(new CraftMapColorCache(this.logger));
         }
         datapackManager = new io.papermc.paper.datapack.PaperDatapackManager(console.getPackRepository()); // Paper
+
+        // Utopia start
+        PropertyPool.loadDefault(console.settings.getProperties(), this);
+        Timer timer = new Timer();
+        timer.schedule(new PerformanceWatchdog(), 0, UtopiaConfig.performanceWatchdogDelay);
+        // Utopia end
     }
 
     public boolean getCommandBlockOverride(String command) {
@@ -389,6 +384,12 @@ public final class CraftServer implements Server {
     private void overrideSpawnLimits() {
         for (SpawnCategory spawnCategory : SpawnCategory.values()) {
             if (CraftSpawnCategory.isValidForLimits(spawnCategory)) {
+                // Utopia start
+                if (UtopiaConfig.configManagementEnabled) {
+                    this.spawnCategoryLimit.put(spawnCategory, PropertyPool.getSpawnCategoryLimit(spawnCategory));
+                    continue;
+                }
+                // Utopia end
                 this.spawnCategoryLimit.put(spawnCategory, this.configuration.getInt(CraftSpawnCategory.getConfigNameSpawnLimit(spawnCategory)));
             }
         }
@@ -694,7 +695,7 @@ public final class CraftServer implements Server {
 
     @Override
     public int getViewDistance() {
-        return this.getProperties().viewDistance;
+        return PropertyPool.viewDistance; // Utopia
     }
 
     @Override
@@ -961,6 +962,7 @@ public final class CraftServer implements Server {
 
         org.spigotmc.SpigotConfig.init((File) console.options.valueOf("spigot-settings")); // Spigot
         this.console.paperConfigurations.reloadConfigs(this.console);
+        ru.bortexel.utopia.UtopiaConfig.init((File) console.options.valueOf("utopia-settings")); // Utopia
         for (ServerLevel world : this.console.getAllLevels()) {
             // world.serverLevelData.setDifficulty(config.difficulty); // Paper - per level difficulty
             world.setSpawnSettings(world.serverLevelData.getDifficulty() != Difficulty.PEACEFUL && config.spawnMonsters, config.spawnAnimals); // Paper - per level difficulty (from MinecraftServer#setDifficulty(ServerLevel, Difficulty, boolean))
@@ -2180,6 +2182,11 @@ public final class CraftServer implements Server {
     }
     public int getSpawnLimitUnsafe(final SpawnCategory spawnCategory) {
         // Paper end
+        // Utopia start
+        if (UtopiaConfig.configManagementEnabled) {
+            return PropertyPool.getSpawnCategoryLimit(spawnCategory);
+        }
+        // Utopia end
         return this.spawnCategoryLimit.getOrDefault(spawnCategory, -1);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index f0b14914438840bd819fa7da8b76f4fcc13704d0..b36c05fd9f2666235d159ea2cb65ece150a32e96 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -135,6 +135,8 @@ import org.bukkit.util.Consumer;
 import org.bukkit.util.RayTraceResult;
 import org.bukkit.util.StructureSearchResult;
 import org.bukkit.util.Vector;
+import ru.bortexel.utopia.UtopiaConfig;
+import ru.bortexel.utopia.management.PropertyPool;
 
 public class CraftWorld extends CraftRegionAccessor implements World {
     public static final int CUSTOM_DIMENSION_OFFSET = 10;
@@ -1817,6 +1819,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return this.getSpawnLimitUnsafe(spawnCategory);
     }
     public final int getSpawnLimitUnsafe(final SpawnCategory spawnCategory) {
+        // Utopia start
+        if (UtopiaConfig.configManagementEnabled) {
+            return PropertyPool.getSpawnCategoryLimit(spawnCategory);
+        }
+        // Utopia end
         int limit = this.spawnCategoryLimit.getOrDefault(spawnCategory, -1);
         if (limit < 0) {
             limit = this.server.getSpawnLimitUnsafe(spawnCategory);
@@ -1830,7 +1837,8 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         Validate.notNull(spawnCategory, "SpawnCategory cannot be null");
         Validate.isTrue(CraftSpawnCategory.isValidForLimits(spawnCategory), "SpawnCategory." + spawnCategory + " are not supported.");
 
-        this.spawnCategoryLimit.put(spawnCategory, limit);
+        if (!UtopiaConfig.configManagementEnabled) // Utopia
+            this.spawnCategoryLimit.put(spawnCategory, limit);
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 5b48047242e20c216b110502935763ea433c55cd..950bf56604037c9e926aa961746a8e5daba73ecf 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -158,6 +158,14 @@ public class Main {
                         .describedAs("Jar file");
                 // Paper end
 
+                // Utopia Start
+                acceptsAll(asList("utopia", "utopia-settings"), "File for utopia settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("utopia.yml"))
+                        .describedAs("Yml file");
+                // Utopia end
+
                 // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
                         .withRequiredArg()
diff --git a/src/main/java/ru/bortexel/utopia/UtopiaConfig.java b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..b6afe711cad47e34280b7b44e363ca770c1494b3
--- /dev/null
+++ b/src/main/java/ru/bortexel/utopia/UtopiaConfig.java
@@ -0,0 +1,131 @@
+package ru.bortexel.utopia;
+
+import com.google.common.base.Throwables;
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+import ru.bortexel.utopia.management.ControlledProperty;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.sql.Time;
+import java.util.HashMap;
+import java.util.List;
+import java.util.logging.Level;
+
+public class UtopiaConfig {
+
+    private static File CONFIG_FILE;
+    private static final String HEADER = "This is the main configuration file for Utopia.";
+    /*========================================================================*/
+    public static YamlConfiguration config;
+    /*========================================================================*/
+
+    public static void init(File configFile) {
+        CONFIG_FILE = configFile;
+        config = new YamlConfiguration();
+
+        try {
+            config.load(CONFIG_FILE);
+        } catch (IOException ignored) {
+        } catch (InvalidConfigurationException ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not load utopia.yml, please correct your syntax errors", ex);
+            throw Throwables.propagate(ex);
+        }
+
+        config.options().header(HEADER);
+        config.options().copyDefaults(true);
+        readConfig();
+    }
+
+    static void readConfig() {
+        for (Method method : UtopiaConfig.class.getDeclaredMethods()) {
+            if (Modifier.isPrivate(method.getModifiers())) {
+                if (method.getParameterTypes().length == 0 && method.getReturnType() == Void.TYPE) {
+                    try {
+                        method.setAccessible(true);
+                        method.invoke(null);
+                    } catch (InvocationTargetException ex) {
+                        throw Throwables.propagate(ex.getCause());
+                    } catch (Exception ex) {
+                        Bukkit.getLogger().log(Level.SEVERE, "Error invoking " + method, ex);
+                    }
+                }
+            }
+        }
+
+        try {
+            config.save(CONFIG_FILE);
+        } catch (IOException ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not save " + CONFIG_FILE, ex);
+        }
+    }
+
+    public static void set(String path, Object val) {
+        config.set(path, val);
+    }
+
+    public static boolean getBoolean(String path, boolean def) {
+        config.addDefault(path, def);
+        return config.getBoolean(path, config.getBoolean(path));
+    }
+
+    public static double getDouble(String path, double def) {
+        config.addDefault(path, def);
+        return config.getDouble(path, config.getDouble(path));
+    }
+
+    public static float getFloat(String path, float def) {
+        // TODO: Figure out why getFloat() always returns the default value.
+        return (float) getDouble(path, (double) def);
+    }
+
+    public static int getInt(String path, int def) {
+        config.addDefault(path, def);
+        return config.getInt(path, config.getInt(path));
+    }
+
+    public static <T> List getList(String path, T def) {
+        config.addDefault(path, def);
+        return (List<T>) config.getList(path, config.getList(path));
+    }
+
+    public static String getString(String path, String def) {
+        config.addDefault(path, def);
+        return config.getString(path, config.getString(path));
+    }
+
+    public static String MANAGEMENT_ROOT = "config-management";
+    public static String MANAGEMENT_PROPERTIES = MANAGEMENT_ROOT + ".properties";
+    public static boolean configManagementEnabled = false;
+    public static Time configManagementEnableFrom = null;
+    public static Time configManagementEnableTill = null;
+    public static int performanceWatchdogDelay = 60000;
+
+    private static void configManagement() {
+        configManagementEnabled = getBoolean(MANAGEMENT_ROOT + ".enable.value", false);
+        if (configManagementEnabled) {
+            loadControlledProperty(new ControlledProperty("view-distance", 6, 10, 10));
+            loadControlledProperty(new ControlledProperty("monster-limit", 5, 70, 10));
+            loadControlledProperty(new ControlledProperty("animal-limit", 2, 15, 7));
+            loadControlledProperty(new ControlledProperty("water-animal-limit", 1, 5, 7));
+            loadControlledProperty(new ControlledProperty("ambient-limit", 2, 15, 7));
+            loadControlledProperty(new ControlledProperty("water-ambient-limit", 2, 20, 5));
+            loadControlledProperty(new ControlledProperty("water-underground-creature-limit", 1, 5, 7));
+            loadControlledProperty(new ControlledProperty("axolotls-limit", 1, 5, 5));
+        }
+
+        configManagementEnableFrom = Time.valueOf(getString(MANAGEMENT_ROOT + ".enable.from", "00:00:00"));
+        configManagementEnableTill = Time.valueOf(getString(MANAGEMENT_ROOT + ".enable.till", "23:59:59"));
+        performanceWatchdogDelay = getInt(MANAGEMENT_ROOT + ".delay", 60000);
+    }
+
+    public static HashMap<String, ControlledProperty> controlledProperties = new HashMap<>();
+
+    private static void loadControlledProperty(ControlledProperty property) {
+        controlledProperties.put(property.getName(), property);
+    }
+}
diff --git a/src/main/java/ru/bortexel/utopia/management/ControlledProperty.java b/src/main/java/ru/bortexel/utopia/management/ControlledProperty.java
new file mode 100644
index 0000000000000000000000000000000000000000..3d00d5837a786fc80955d1c814321b74a8287571
--- /dev/null
+++ b/src/main/java/ru/bortexel/utopia/management/ControlledProperty.java
@@ -0,0 +1,36 @@
+package ru.bortexel.utopia.management;
+
+import org.bukkit.configuration.ConfigurationSection;
+import ru.bortexel.utopia.UtopiaConfig;
+
+public class ControlledProperty {
+    public double min;
+    public double max;
+    public int weight;
+    private final String name;
+
+    public ControlledProperty(ConfigurationSection configuration) {
+        this.min = configuration.getDouble("min");
+        this.max = configuration.getDouble("max");
+        this.weight = configuration.getInt("weight");
+        this.name = configuration.getCurrentPath();
+    }
+
+    public ControlledProperty(String name, double defMin, double defMax, int defWeight) {
+        String path = UtopiaConfig.MANAGEMENT_PROPERTIES + "." + name + ".";
+        this.min = UtopiaConfig.getDouble(path + "min", defMin);
+        this.max = UtopiaConfig.getDouble(path + "max", defMax);
+        this.weight = UtopiaConfig.getInt(path + "weight", defWeight);
+        this.name = name;
+    }
+
+    public double getValue(double lagRate) {
+        // TODO: Add weight support
+        double value = this.min + (this.max - this.min) * lagRate / 100;
+        return Math.max(this.min, Math.min(this.max, value));
+    }
+
+    public String getName() {
+        return name;
+    }
+}
diff --git a/src/main/java/ru/bortexel/utopia/management/PerformanceWatchdog.java b/src/main/java/ru/bortexel/utopia/management/PerformanceWatchdog.java
new file mode 100644
index 0000000000000000000000000000000000000000..5d007b512bdfa58b46bd8c9c692a94db36fdceee
--- /dev/null
+++ b/src/main/java/ru/bortexel/utopia/management/PerformanceWatchdog.java
@@ -0,0 +1,50 @@
+package ru.bortexel.utopia.management;
+
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.Bukkit;
+import ru.bortexel.utopia.UtopiaConfig;
+
+import java.sql.Time;
+import java.time.LocalTime;
+import java.util.TimerTask;
+
+public class PerformanceWatchdog extends TimerTask {
+    @Override
+    public void run() {
+        Time now = Time.valueOf(LocalTime.now());
+        if (!UtopiaConfig.configManagementEnabled) return;
+        if (UtopiaConfig.configManagementEnableFrom.after(now) || UtopiaConfig.configManagementEnableTill.before(now))
+            return;
+
+        double currentTPS = Math.max(0, Bukkit.getTPS()[0] - 10);
+        double maxTPS = MinecraftServer.TPS - 10;
+        double rate = currentTPS / maxTPS * 100;
+
+        try {
+            ControlledProperty viewDistance = UtopiaConfig.controlledProperties.get("view-distance");
+            ControlledProperty monsterSpawnLimit = UtopiaConfig.controlledProperties.get("monster-limit");
+            ControlledProperty animalSpawnLimit = UtopiaConfig.controlledProperties.get("animal-limit");
+            ControlledProperty waterAnimalSpawnLimit = UtopiaConfig.controlledProperties.get("water-animal-limit");
+            ControlledProperty ambientSpawnLimit = UtopiaConfig.controlledProperties.get("ambient-limit");
+            ControlledProperty waterAmbientSpawnLimit = UtopiaConfig.controlledProperties.get("water-ambient-limit");
+            ControlledProperty waterUndergroundCreatureSpawnLimit = UtopiaConfig.controlledProperties.get("water-ambient-underground-creature-limit");
+
+            if (viewDistance != null)
+                PropertyPool.viewDistance = (int) Math.round(viewDistance.getValue(rate));
+            if (monsterSpawnLimit != null)
+                PropertyPool.monsterSpawnLimit = (int) Math.round(monsterSpawnLimit.getValue(rate));
+            if (animalSpawnLimit != null)
+                PropertyPool.animalSpawnLimit = (int) Math.round(animalSpawnLimit.getValue(rate));
+            if (waterAnimalSpawnLimit != null)
+                PropertyPool.waterAnimalSpawnLimit = (int) Math.round(waterAnimalSpawnLimit.getValue(rate));
+            if (ambientSpawnLimit != null)
+                PropertyPool.ambientSpawnLimit = (int) Math.round(ambientSpawnLimit.getValue(rate));
+            if (waterAmbientSpawnLimit != null)
+                PropertyPool.waterAmbientSpawnLimit = (int) Math.round(waterAmbientSpawnLimit.getValue(rate));
+            if (waterUndergroundCreatureSpawnLimit != null)
+                PropertyPool.waterUndergroundCreatureSpawnLimit = (int) Math.round(waterUndergroundCreatureSpawnLimit.getValue(rate));
+        } catch (Exception e) {
+            Bukkit.getLogger().warning("Performance Watchdog failure: " + e.getLocalizedMessage());
+        }
+    }
+}
diff --git a/src/main/java/ru/bortexel/utopia/management/PropertyPool.java b/src/main/java/ru/bortexel/utopia/management/PropertyPool.java
new file mode 100644
index 0000000000000000000000000000000000000000..d2ae3d8f78fdf3892975e8d95d237a56ea0448ec
--- /dev/null
+++ b/src/main/java/ru/bortexel/utopia/management/PropertyPool.java
@@ -0,0 +1,33 @@
+package ru.bortexel.utopia.management;
+
+import net.minecraft.server.dedicated.DedicatedServerProperties;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.entity.SpawnCategory;
+
+public class PropertyPool {
+    public static Integer viewDistance = 10;
+    public static Integer monsterSpawnLimit = 70;
+    public static Integer animalSpawnLimit = 15;
+    public static Integer waterAnimalSpawnLimit = 5;
+    public static Integer ambientSpawnLimit = 15;
+    public static Integer waterAmbientSpawnLimit = 20;
+    public static Integer waterUndergroundCreatureSpawnLimit = 5;
+    public static Integer axolotlsSpawnLimit = 5;
+
+    public static void loadDefault(DedicatedServerProperties properties, CraftServer server) {
+        viewDistance = properties.viewDistance;
+    }
+
+    public static int getSpawnCategoryLimit(SpawnCategory category) {
+        return switch (category) {
+            case MONSTER -> monsterSpawnLimit;
+            case ANIMAL -> animalSpawnLimit;
+            case WATER_ANIMAL -> waterAnimalSpawnLimit;
+            case WATER_AMBIENT -> waterAmbientSpawnLimit;
+            case WATER_UNDERGROUND_CREATURE -> waterUndergroundCreatureSpawnLimit;
+            case AMBIENT -> ambientSpawnLimit;
+            case AXOLOTL -> axolotlsSpawnLimit;
+            default -> throw new UnsupportedOperationException("Unknown spawn category " + category);
+        };
+    }
+}
