From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RuscalWorld <me@ruscalworld.ru>
Date: Fri, 9 Jul 2021 16:02:10 +0300
Subject: [PATCH] Make performance watchdog run right after initialization


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 2d2c6dd49bbeaf25ae0e934a728bfeb9346c6760..1403885a85e2636617a59adda3e25e9b514f662a 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -375,7 +375,7 @@ public final class CraftServer implements Server {
         // Utopia start
         PropertyPool.loadDefault(console.settings.getProperties(), this);
         Timer timer = new Timer();
-        timer.schedule(new PerformanceWatchdog(), UtopiaConfig.performanceWatchdogDelay / 2, UtopiaConfig.performanceWatchdogDelay);
+        timer.schedule(new PerformanceWatchdog(), 0, UtopiaConfig.performanceWatchdogDelay);
         // Utopia end
     }
 
